src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -22,40 +22,159 @@ import { first } from 'rxjs/operators';
 import {
   SavedObjectsClient,
   SavedObjectsSchema,
-  SavedObjectsRepository,
-  SavedObjectsSerializer,
   SavedObjectsClientProvider,
   ISavedObjectsClientProvider,
+  SavedObjectsClientProviderOptions,
 } from './';
-import { getRootPropertiesObjects } from './mappings';
 import { KibanaMigrator, IKibanaMigrator } from './migrations';
 import { CoreContext } from '../core_context';
-import { LegacyServiceSetup } from '../legacy/legacy_service';
-import { ElasticsearchServiceSetup } from '../elasticsearch';
+import { LegacyServiceDiscoverPlugins } from '../legacy/legacy_service';
+import { ElasticsearchServiceSetup, APICaller } from '../elasticsearch';
 import { KibanaConfigType } from '../kibana_config';
-import { retryCallCluster, migrationsRetryCallCluster } from '../elasticsearch/retry_call_cluster';
+import { migrationsRetryCallCluster } from '../elasticsearch/retry_call_cluster';
 import { SavedObjectsConfigType } from './saved_objects_config';
 import { KibanaRequest } from '../http';
+import { SavedObjectsClientContract } from './types';
+import { ISavedObjectsRepository } from './service/lib/repository';
+import {
+  SavedObjectsClientFactory,
+  SavedObjectsClientWrapperFactory,
+} from './service/lib/scoped_client_provider';
+import { createRepository } from './service/lib/create_repository';
 import { Logger } from '..';
 
 /**
+ * Saved Objects is Kibana's data persisentence mechanism allowing plugins to
+ * use Elasticsearch for storing and querying state. The
+ * SavedObjectsServiceSetup API exposes methods for creating and registering
+ * Saved Object client wrappers.
+ *
+ * @remarks
+ * Note: The Saved Object setup API's should only be used for creating and
+ * registering client wrappers. Constructing a Saved Objects client or
+ * repository for use within your own plugin won't have any of the registered
+ * wrappers applied and is considered an anti-pattern. Use the Saved Objects
+ * client from the
+ * {@link SavedObjectsServiceStart | SavedObjectsServiceStart#scopedClient }
+ * method or the {@link RequestHandlerContext | route handler context} instead.
+ *
+ * When plugins access the Saved Objects client, a new client is created using
+ * the factory provided to `setClientFactory` and wrapped by all wrappers
+ * registered through `addClientWrapper`. To create a factory or wrapper,
+ * plugins will have to construct a Saved Objects client. First create a
+ * repository by calling `scopedRepository` or `internalRepository` and then
+ * use this repository as the argument to the {@link SavedObjectsClient}
```

Comment:
Honest question, is the `setClientFactory` isolated per-plugin, or is this the responsibility or a single plugin to define it for everyone? What is the actual expected usage of this setter?

Reply:
These extension points to Saved Objects are there to support the x-pack security plugins: security, encrypted saved objects and spaces and they apply globally, to all plugins that use the saved objects client. This makes the documentation hard to write, cause it's there to solve a very specific problem.

There's broad consensus that this factory and the client wrappers aren't the most elegant architecture, but the way it worked in legacy did the job so we won't try to improve this until after 8.0.0. The security team might have some upcoming requirements that require something more like middleware than just composing clients over each other, but there's no concrete plans yet.

Reply:
Not sure if we're there yet, but if we are **sure** that this pattern will be replaced, we should mark the `setClientFactory` and `addClientWrapper` methods `@deprecated`.

Reply:
A public API that allows only to be called once sure seems like fragile :) But this will do the job until proper revision.